Parser/pegen.c

```diff
@@ -456,10 +456,12 @@ _PyPegen_raise_error_known_location(Parser *p, PyObject *errtype,
         goto error;
     }
 
+    int uses_utf8_codec = (!p->tok->encoding || strcmp(p->tok->encoding, "utf-8") == 0);
+
     if (p->tok->fp_interactive) {
         error_line = get_error_line(p, lineno);
     }
-    else if (p->start_rule == Py_file_input) {
+    else if (uses_utf8_codec && p->start_rule == Py_file_input) {
         error_line = PyErr_ProgramTextObject(p->tok->filename, (int) lineno);
```

Comment:
Hmm, `PyErr_ProgramTextObject` is part of the public API. I wonder if we should document somewhere that it should only be used with source files encoded with utf-8 or else it can result incorrect strings.

Reply:
Is part of the public API, but is not documented, so I prefer to leave it as it as there is no guarantees here.